### PR TITLE
[llvm] Add support for running tests as root

### PR DESCRIPTION
**Pull request by tstellar**
There are a few test that check access permissions, so they need to be disabled when running the tests as root.

The most common use case for running tests as root is inside of a container.  GitHub Actions, for example, only supports running the root user inside of containers, so this change is necessary in order to run the tests inside of a container running in the GitHub Actions environment.

### CODE REVIEW

**Review by boomanaiden154 — APPROVED** [review]
LGTM. Please wait for @MaskRay / others to chime in though.

Thanks for putting this together. Given the CI constraints, I think there's a really compelling case for adding this in.

**Review by jh7370 — APPROVED** [review]
LGTM too.

**Review by RoboTux — APPROVED** [review]
LGTM. This would have been very useful in my previous job, I'm sure they'll be delighted when they merge this in. Thanks!

**Comment by arichardson** [comment]
How about testing for the ACL bypassing behaviour instead of checking for username == root?

```sh
$ sudo python3 -c '
import tempfile
from pathlib import Path

with tempfile.TemporaryDirectory() as td:
    Path(td, "test.txt").write_text("test")
    Path(td, "test.txt").chmod(0o100)
    print(Path(td, "test.txt").read_text())
'

test


python3 -c '
import tempfile
from pathlib import Path

with tempfile.TemporaryDirectory() as td:
    Path(td, "test.txt").write_text("test")
    Path(td, "test.txt").chmod(0o100)
    print(Path(td, "test.txt").read_text())
'
Traceback (most recent call last):
  File "<string>", line 8, in <module>
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/pathlib.py", line 1256, in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/pathlib.py", line 1242, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/pathlib.py", line 1110, in _opener
    return self._accessor.open(self, flags, mode)
PermissionError: [Errno 13] Permission denied: '/var/folders/_5/ltg5zrm5379_yh38z_j2vvgr00w_dz/T/tmpu5g01tt8/test.txt'
```

File: llvm/utils/lit/lit/llvm/config.py [hunk 1]
```diff
@@ -13,6 +14,18 @@
 lit_path_displayed = False
 
 
+def user_is_root():
+    # getpass.getuser() can throw an exception in some cases:
+    # See https://github.com/python/cpython/issues/76912
+    try:
+        if getpass.getuser() == "root":
```
Comment (MaskRay) [inline llvm/utils/lit/lit/llvm/config.py:21]: *NIX systems can use `os.getuid == 0`

**Review by MaskRay — APPROVED** [review]

**Comment by tstellar** [comment]
Any thoughts on using os.getuid() vs an explicit test for the ACL bypassing behavior?

**Comment by tstellar** [comment]
> How about testing for the ACL bypassing behaviour instead of checking for username == root?

@arichardson  How portable would this check be?  Are there any scenarios where this check would return a different result than checking username == root?

**Comment by RoboTux** [comment]
> > How about testing for the ACL bypassing behaviour instead of checking for username == root?
> 
> @arichardson How portable would this check be? Are there any scenarios where this check would return a different result than checking username == root?

binary with setuid bit will have getuser() return the user that ran the binary but the binary will behave as root.

**Comment by arichardson** [comment]
> > > How about testing for the ACL bypassing behaviour instead of checking for username == root?
> > 
> > 
> > @arichardson How portable would this check be? Are there any scenarios where this check would return a different result than checking username == root?
> 
> binary with setuid bit will have getuser() return the user that ran the binary but the binary will behave as root.

But we are checking for the permissions that the python interpreter + spawned binaries have, so setuid binaries don't matter. If llvm-ar was setuid we would have a whole new set of problems regardless of current user privileges.

I don't feel strongly which check is used I just thought checking whether the behaviour that makes the test fail is present is slightly cleaner than doing so indirectly via getuid==0. The main problem I can think of with the ACL bypassing check is that Windows might behave weirdly but the tests are already disabled on Windows.

**Comment by RoboTux** [comment]
> > > > How about testing for the ACL bypassing behaviour instead of checking for username == root?
> > > 
> > > 
> > > @arichardson How portable would this check be? Are there any scenarios where this check would return a different result than checking username == root?
> > 
> > 
> > binary with setuid bit will have getuser() return the user that ran the binary but the binary will behave as root.
> 
> But we are checking for the permissions that the python interpreter + spawned binaries have, so setuid binaries don't matter. If llvm-ar was setuid we would have a whole new set of problems regardless of current user privileges.

All we need is for lit to be setuid for the check to return false despite everything being run with root privilege, right? And yes it might lead to privilege issue but so does running lit as root. That said I'm not arguing for a change, I didn't make the ACL comment. As pointed out this is quite useful for running things in containers and if someone goes to the trouble of having the container have a non root user but still use setuid it's looking for trouble. Not sure about ACL but I would assume it's similar: bad idea outside of a container because lit can run anything, all one need is create a test file with the right RUN line.

In short, I'm personally happy with the patch as it is.

My 2 cents.

**Comment by tstellar** [comment]
Thanks for the feedback.  I would prefer to keep the patch as-is.  The reason I don't want to use os.getuid() is that it is only available on Unix systems, and I don't want to have to add extra code to do OS detection.  I also don't want to do a specific ACL test, because creating temporary files is adding another point of failure to lit start up.  Since this check is only needed by 4 tests, I want to make it as simple as possible and reduce the risk that it will impact users running as non-root, which is the most common case.

**Comment by arichardson** [comment]
> Thanks for the feedback. I would prefer to keep the patch as-is. The reason I don't want to use os.getuid() is that it is only available on Unix systems, and I don't want to have to add extra code to do OS detection. I also don't want to do a specific ACL test, because creating temporary files is adding another point of failure to lit start up. Since this check is only needed by 4 tests, I want to make it as simple as possible and reduce the risk that it will impact users running as non-root, which is the most common case.

That sounds fine to me. I think using os.getuid instead of getpass is still possible since it's already wrapped inside a bare try-except so that should be safe and avoids an extra import?

**Comment by tstellar** [comment]
> That sounds fine to me. I think using os.getuid instead of getpass is still possible since it's already wrapped inside a bare try-except so that should be safe and avoids an extra import?

I tested this and it works, so I've updated the patch to use os.getuid now.

**Review by RoboTux — APPROVED** [review]
Still LGTM
